### ISSUE TITLE
[2.0.3] `getSemanticHTML` is broken

### ISSUE THREAD

**Issue by boris-petrov** [slab/quill#4509]
**Steps for Reproduction**

1. [Reproduction](https://quilljs.com/playground/snow#codeN4Ig9AlgdgJgpgDwHQAsAuBbANiAXCAHhggDcACCGAXgB0Q5i0B7AJzoD4BDMtRNMgO4Q0KMhgCuWNBAAOWOGQDOMzgGM4igmGIl2NKPoKLVLWf0UtVtcNHjIAVoo5bjpmWnYgANDdiIkjnggqkxQivwAjuIQWFhkVGRQcAJkAIrRsQAUAOQAxAzCrNleZMD6ZGJMMJIauKXlFTxMTFgARpwsdQDaDY1kXcBkKHCc8J39AIwlAEwlAGacWIpwALpkAL4rXr2NXdmtLTDFZNnCixCqx9nifixY0HDZWzsVexAYnADmjyXZIfAAWlaWCYqgA1k9tlA+s9oRsoRU5Go4ChDnBxtkAMJMDAyJjLMicaFwGQXJDk4oNERwDBwOrZRRQJgCY5gMBkVgnVriVrAx76dYASgA3Pp9CEwi04EgQZ9MlMyFEYlgkN80ABlGlE6SqAASABUALIAGUygsFIHWQA)
2. Check the console

**Expected behavior**:
The text in the editor is `a text with multiple spaces`. `getSemanticHTML` returns `<p>a&nbsp;text&nbsp;with&nbsp;multiple&nbsp;spaces</p>`. Which is wrong - the semantics of the text has changed.

**Actual behavior**:
See above.

**Platforms**:
Any.

**Version**:
2.0.3

**Comment by islandmyst** [comment]
Introduced wtih Commit 07b68c9.

The end result of replacing all spaces with `&nbsp;` (Non Breaking Space) is that the words no-long wrap at word boundaries, and instead overflow.

[Example](https://quilljs.com/playground/snow#codeN4Ig9AlgdgJgpgDwHQAsAuBbANiAXCAHhggDcACCGAXgB0Q5i0B7AJzoD4BZAVyzQizQ4ZACqI0SMgDkmaYWhTCA7qxhklLAIYAHbdADmZJlDILh+uGgDKcDJqj8AxgAkRnADIAKAJRkWcWDgWA0l3VlsKbQBnbgwyGCYsVjIoiDQyTQxLABoyR2MouEc5NG4WDOI9KMcDMjhBNFzCtQS6iG4ojCY1OQxtZOga4hhuBzJudKxNACNwuvSEpP8yO30oTQzBAEduTUkAVXSAiDjNNQxoE7ISY8zcnYgosigmKLQWbjVEIJq0TX5jOMsFMMPkyFNZsEnlBHhBxulNNtuBBtHUEHUNvkMF1WvkoIUdv8kAQwMQSOwaFBKQRqsFtOkoixHLRwNB4MgAFZRDgk2kotAUqlQIikCjUOgWay2exOVweAD6+W0AE8eaTSOwQNlWYFOVE8CA8W8yA9gWQqM84EoyABFZHAzwAcgAxAw0qxHblgJSyCturw4FFcGRvSZfaYmIlppoWMGANo+8O+uPAMiKM5BeMARlyACZcgAzRGFAC6ZAAviXsomk3HHbMsDBPWRHWlERBHM3HaN4CxBFA4I6qzXw3WTpoLF38vAALTTJKOADWQ+rYd9w7D5dXvu0U0ccBQiV7wcdAGEmH1XsJ7HU9I4kA-PYmzFkT1EXkpm2AwEZyvXuNM86DpS5beAA3JSlKFGgIgnHATATJ4PjmuwIaJgkjixAEEiSgAolgtjYQAQsqACSMBOpKNh2A4HZyu4ipMCqjreEg0ADiw9Hmia9pYEgVHSrRLhuF44EgbkWYAAzSd4kFQEaiRwEgST6J4OY8QIfECTRsoiT43ggOWQA)

An alternative implementation of editor.js converHTML() that preserves whitespace might replace all but one consecutive space character with &nbsp; be:
```JavaScript
  if (blot instanceof TextBlot) {
    const escapedText = escapeText(blot.value().slice(index, index + length));
    return escapedText.replaceAll(/  +/g, (match) => "&nbsp;".repeat(match.length - 1) + " ")
  }
```

A work around (which replaces the last consecutive `&nbsp;` with a space) until a fix is implemented is:
```JavaScript
quill.getSemanticHTML().replaceAll(/((?:&nbsp;)*)&nbsp;/g, '$1 ')
```

**Comment by frederikhors** [comment]
I think this is bad. Sometimes I need to write `&nbsp;` literally in the editor.

Can we revert to the amazing `getSemanticHTML()` method of the 2.0.2 please?

**Comment by Sergiobop** [comment]
I think @luin can help us here

Thanks

**Comment by cami-dev** [comment]
We have the same issue please revert this change. Don't want to save nbps to the database.

**Comment by TheSeg** [comment]
Spent the entire day only to discover this is a known and unaddressed bug.

The problem with all spaces becoming `&nbsp;` is that it completely breaks pervious word wrapping. It's a non-breaking space, thus all instances that would otherwise line wrap suddenly won't. For a WYSIWYG editor, what you see is currently _not_ what you get.

**Comment by Anthonyyp** [comment]
commenting to follow, hoping to see an update here.

**Comment by michal-laskowski** [comment]
IMHO this is critical bug

**Comment by Hapanmuffinssi** [comment]
> IMHO this is critical bug

Yes. Very much so. For example, ngx-quill (https://github.com/KillerCodeMonkey/ngx-quill) also suffers from this. If the editor is used as part of Angular reactive forms, the value of the form control has these added &nbsp; characters after user has finished typing something with whitespaces.

**Comment by JonasTriki** [comment]
Any update on this @luin?

**Comment by KyeRussell** [comment]
The fact that this hasn't been addressed yet has me concerned that I'm not...using Quill correctly, or something. This respectfully seems like an incredibly critical bug. Is there something I'm missing?

**Comment by VahidN** [comment]
>  Is there something I'm missing?

Yes, it's under the name of `Slab` now, which means soon or late it will become a commercial product (based on my experience throughout these years!).

**Comment by DirkLachowski** [comment]
@VahidN 

> Yes, it's under the name of Slab now, which means soon or late it will become a commercial product (based on my experience throughout these years!).

Not too sure about this. Biggest pro of Quill is that it's free, not that it's feature complete. If I have to pay for it, there's much better payed alternatives, eg. Freola. I'm rather concerned that Slab is abandoning it. There's 400+ open issues, 60 open PRs and not really much to be seen from the maintainers

**Comment by pahen** [comment]
Please, just say you have abonded it @jhchen and @luin so we can move forward and find a solution for the state Quill is in now.

**Comment by Sergiobop** [comment]
> Please, just say you have abonded it [@jhchen](https://github.com/jhchen) and [@luin](https://github.com/luin) so we can move forward and find a solution for the state Quill is in now.

Don't be dramatic, just stay on version 2.0.2 for now, until they can fix it.

Before version 2, the repository was inactive for over 4 years. It's only been 2-3 months since version 2.0.3

**Comment by frederikhors** [comment]
Yeah. Not another drama please.

**Comment by boosh** [comment]
Jeez this thread is a joke. Just do this in your code: `input.replace("&nbsp;", " ")`

**Comment by DirkLachowski** [comment]
> The fact that this hasn't been addressed yet has me concerned that I'm not...using Quill correctly, or something. This respectfully seems like an incredibly critical bug. Is there something I'm missing?

Probably you are. Most comments complain that this bug has them save `&nbsp;`s to the DB. If that’s also your problem then yes, you’re using it wrong. You’re supposed to save the delta format, not something rendered. If you need to send html from the ui to your backend, Quill maybe isn’t the right editor for your requirements. If you need html for whatever reason, parse the delta. That’s the usp of Quill, a view agnostic storage format.

**Comment by frederikhors** [comment]
> Probably you are.

Or, since there are so many of us here, our way of using Quill is not wrong either. 🥰

**Comment by DirkLachowski** [comment]
@frederikhors 

> Or, [@DirkLachowski](https://github.com/DirkLachowski), since there are so many of us here, our way of using Quill is not wrong either. 🥰

Well, "correct" is literally there in @KyeRussell 's [question](https://github.com/slab/quill/issues/4509#issuecomment-2612664830), I just reused that wording when answering it. And I beg to disagree, there's an intended way to use Quill. 

From the Delta readme:

> Deltas can describe any rich text document, includes all text and formatting information, without the ambiguity and complexity of HTML.
(from https://github.com/slab/delta ). 

The real problem is that there are not that much Delta to whatever render tools, so people resort to Quill as a Delta renderer. That works sometimes, but it's not a great idea.

**Comment by raymondelferink** [comment]
@islandmyst Your solution works for me. Thank you for posting that!
Did you make a pull request for that fix?

**Comment by TheSeg** [comment]
> [@frederikhors](https://github.com/frederikhors)
> Well, "correct" is literally there in [@KyeRussell](https://github.com/KyeRussell) 's [question](https://github.com/slab/quill/issues/4509#issuecomment-2612664830), I just reused that wording when answering it. And I beg to disagree, there's an intended way to use Quill.

If that is the case, then why is `getSemanticHTML()` a [supported function](https://quilljs.com/docs/api#getsemantichtml)? Emphasis is mine:

> Get the HTML representation of the editor contents. This method is useful for exporting the contents of the editor **_in a format that can be used in other applications_**.

An end user who presented with wrapping lines in an editor intends to have lines wrap. Swapping for `&nbsp;` destroys the intentions of the end user's line wrapping. It is simply the incorrect interpretation of the end user's intentions.

If the fix is using the Delta system, then `getSemanticHTML()` is an unsupported function as it doesn't generate accurate HTML. This would also change the scope of Quill and explicitly only compatible with Delta.

If the fix is to revert 07b68c98b3bb61e6a6efb369a76f44a16b278517 and allow `getSemanticHTML()` to output the intended HTML, then we would have the expected results.

**Comment by DirkLachowski** [comment]
@TheSeg Tell that the maintainers, not me. If you think that that function needs a fix, you can open a PR and we'll see if it will be merged. That said, `getSemanticHTML` has been added with v2, so we can have a discussion if adding `&nbsp;` is an error (most likely it is), but it's nowhere defined what the correct output is. Btw, this looks a lot like a private function that just got used a lot by consumers and then got promoted to the public API https://github.com/slab/quill/issues/3992. So, maybe it's just that the maintainers need it right that way in their products. 

That said: Using the Delta format as a representation for content **and** changes is just a big read flag showing lack of upfront design. So, there's maybe more errors to come and it's time to move on. I'll do so.

**Comment by stashaway** [comment]
I opened a pull request with a fix. Hopefully it will get approved.

**Comment by frederikhors** [comment]
The PR is this: https://github.com/slab/quill/pull/4587.

**Comment by MaciejDabek** [comment]
I created an alternative fix (PR #4598) that introduce options to getSemanticHTML, so it can be used as it is (with `&nbsp;`) for internal coping and pasting in quill. Setting `preserveWhitespace` option to _true_ will preserve whitespace for all of us needing it.

**Comment by helen-andronova** [comment]
Any updates?

**Comment by jannes-io** [comment]
Yikes.. breaking backwards compatibility in a patch version... Guess the one take-away we got from this is to fix quill to whatever version works and leave it there until we hit critical vulnerabilities...

**Comment by michal-laskowski** [comment]
this must be some psyop. trust no patch.

**Comment by raymondelferink** [comment]
I think a patch should be based on the suggested fix proposed by @islandmyst (https://github.com/slab/quill/issues/4509#issuecomment-2521334016)
That means adding multiple spaces in the editor still results in multiple spaces in the rendered html, without the crippling absence of actual spaces.
I'm traveling right now, but if no one has time to make a pull-request, then I will make one after the weekend:-)

**Comment by oardi** [comment]
We weren't aware that a bug had been open here for several months.
This has resulted in an unsightly number of non breakable characters in the database.
It would be great if the bug could be fixed finally.

**Comment by makcyd** [comment]
Since the issue is twofold (some need &nbsp; others regular spaces - as we do) maybe it's worth adding optional parameter for `getSemanticHTML` function, that's passed further to [convertHTML](https://github.com/slab/quill/blob/ebe16ca24724ac4f52505628ac2c4934f0a98b85/packages/quill/src/core/editor.ts#L363)?

Say, `getSemanticHTML(index?, length?, replaceSpaces?)`

**Comment by dimitriaguera** [comment]
Hello here,

We are also affected, and using ngx-quill  (https://github.com/KillerCodeMonkey/ngx-quill) so cannot easily revert to 2.0.2 quill version.

Using temporary workaround while waiting for a fix :

On template
```html
<quill-editor (onEditorCreated)="onEditorCreate($event)">
```

On component
```javascript
  onEditorCreated(quill) {
    quill.legacyGetSemanticHTML = quill.getSemanticHTML;
    quill.getSemanticHTML = (a: number, b: number) =>
      quill
        .legacyGetSemanticHTML(a, b)
        .replaceAll(/((?:&nbsp;)*)&nbsp;/g, '$1 ');
  }
```
(Thanks @islandmyst for regex)

Not perfect, but no perfect but make the job.

**Comment by manoj12dsynergy** [comment]
https://github.com/slab/quill/issues/4535#issuecomment-2874963898 This is the solution that worked for me.

**Comment by marsc** [comment]
Any updates on this issue?

**Comment by NateRadebaugh** [comment]
To work around [this regression commit](https://github.com/slab/quill/commit/07b68c98b3bb61e6a6efb369a76f44a16b278517), you can use [`patch-package`](https://www.npmjs.com/package/patch-package)

Specifically, editing `\node_modules\quill\core\editor.js` in `convertHTML` around line `:298`

```diff
diff --git a/node_modules/quill/core/editor.js b/node_modules/quill/core/editor.js
index 07fd544..429968f 100644
--- a/node_modules/quill/core/editor.js
+++ b/node_modules/quill/core/editor.js
@@ -295,7 +295,7 @@ function convertHTML(blot, index, length) {
   }
   if (blot instanceof TextBlot) {
     const escapedText = escapeText(blot.value().slice(index, index + length));
-    return escapedText.replaceAll(' ', '&nbsp;');
+    return escapedText;
   }
   if (blot instanceof ParentBlot) {
     // TODO fix API
```

**Comment by pokonski** [comment]
This bug basically makes the editor unusable...

**Comment by frederikhors** [comment]
Yeah, what happened to quill's team?

**Comment by jannes-io** [comment]
@frederikhors 

> Yeah, what happened to quill's team?

Seems like Slab took a private fork of this repo and use that for their own use-case instead of adding improvements and fixes to this repo. This is evident if you look at this #4672 which I think was supposed to target their private fork instead of this public repo.

Furthermore, if you look at that PR, and you go to [`core/editor.ts`](https://github.com/slab/quill/pull/4672/files#diff-5c90999b27137e1309b91cf9694757bdcb82f67ca870d013d6e1a983cb70d491), you can actually see that in their private version, they've fixed this bug and removed to `replaceAll`.

Technically since this accidental PR was added to this repo, all of its code is now under the same license as this repo. So feel free to grab that PR, merge it into your own fork of Quill, and run with that.

So draw your own conclusions if you will, but seems to me that this public version of the repo is just abandoned, and they have these public repos for marketing and not to actually provide any useful value to the open-source community.

**Comment by targos** [comment]
Are you sure about that? It looks like https://github.com/slab/quill/pull/4672 was opened by someone from a company called Miro, that doesn't seem related to Slab.

**Comment by jannes-io** [comment]
> Are you sure about that? It looks like [#4672](https://github.com/slab/quill/pull/4672) was opened by someone from a company called Miro, that doesn't seem related to Slab.

You're correct. My bad. But my bet is that slab is doing the same thing either way...

The fix is still the same, you can grab Miro App's fork (#4672) now that has this fix, plus many other fixes, since they added it to a public repo 😆

**Comment by frederikhors** [comment]
Yeah but why should the abandon this repo?

**Comment by diabhoil** [comment]
Any update on this?

**Comment by Ibessas** [comment]
No official fix?

**Comment by serhumanos** [comment]
back to v2.02 for now, that version works fine

**Comment by gazpachoking** [comment]
I think an easy fix which would work for everyone is: Instead of replacing all spaces with non breaking spaces, replace two spaces in a row with a non-breaking space followed by a space. This would mean single spaces remain the same, but repeated spaces don't become insignificant like they normally would be in html. (And most importantly, allows text to be broken across lines still.)
i.e.
```diff
- return escapedText.replaceAll(' ', '&nbsp;');
+ return escapedText.replaceAll('  ', '&nbsp; ');
```

**Comment by alextenb** [comment]
Is this library maintained anymore?

**Comment by katywings** [comment]
If you need a temporary drop-in replacement you can use:
https://www.npmjs.com/package/@lufrai/quill

Apart from a fix for this issue it does one thing differently:
it renders the toolbar of the bubble theme with https://floating-ui.com/.

**Comment by esetnik** [comment]
We hit this in production on a CMS app (LMS course descriptions). Adding a small data point and a self-contained reproduction in case it's useful when prioritizing.

### Reproduction (Vitest, jsdom)

```ts
import Quill from 'quill';
import { describe, expect, it } from 'vitest';

function createEditorWith(text: string): Quill {
  const container = document.createElement('div');
  document.body.appendChild(container);
  const quill = new Quill(container);
  quill.setText(text);

  return quill;
}

describe('Quill 2.0.3 nbsp regression (slab/quill#4509)', () => {
  it('preserves regular spaces between words on getSemanticHTML', () => {
    const quill = createEditorWith('a text with multiple spaces\n');

    const html = quill.getSemanticHTML();

    expect(html).not.toMatch(/&nbsp;/);
    expect(html).toBe('<p>a text with multiple spaces</p>');
  });

  it('does not insert &nbsp; for short typed descriptions', () => {
    const quill = createEditorWith('Welcome to the course!\n');

    const html = quill.getSemanticHTML();

    expect(html).not.toMatch(/&nbsp;/);
    expect(html).toBe('<p>Welcome to the course!</p>');
  });
});
```

On stock `quill@2.0.3` both assertions fail with `<p>Welcome&nbsp;to&nbsp;the&nbsp;course!</p>`. Pinning to `quill@2.0.2` or reverting the `replaceAll(' ', '&nbsp;')` line introduced in 07b68c9 makes them pass.

### Real-world impact

- 447 records in our production DB ended up with `&nbsp;` between every word over the ~6 weeks since we upgraded to 2.0.3 (via `react-quill-new@3.8.3`). Worst single record: ~6,400 nbsp entities.
- The serializer change is silent until the rendered paragraph (a) needs to wrap and (b) lives inside a CSS-Grid track sized to `min-content`. When both are true, the unbreakable nbsp-joined line forces the track to expand to its `min-content`, which on our pages pushes adjacent grid items (a video player in one case, a class image in another) off screen.
- It also breaks plain-text consumers — the saved value is no longer the user's actual text.

### Workaround we're shipping

Reverting the `replaceAll(' ', '&nbsp;')` line via `pnpm patch quill@2.0.3`. PR for context: https://github.com/ProdigyEMS/prodigy/pull/12628 (the test above is in `frontend/src/Components/RichText.test.ts`).

We understand the original change in #4502 was trying to preserve visible whitespace that HTML's default would collapse, and we don't have a fully-baked proposal for the right replacement — narrowing to "only `&nbsp;` the spaces that would actually collapse" (consecutive runs, leading/trailing inside inline elements) is presumably the intent but is non-trivial to get right without breaking other round-trip cases. Just sharing data that the current behavior has measurable downstream impact.

[Written by Claude]
